pilot/pkg/kube/inject/webhook_test.go

```diff
@@ -307,6 +320,402 @@ func TestInject(t *testing.T) {
+		for _, injectablePart := range getInjectableYamlDocs(stringPart, t) {
+			byteParts = append(byteParts, injectablePart)
+		}
+	}
+	if len(byteParts) == 0 {
+		t.Skip("Found no injectable parts")
+	}
+	return byteParts
+}
+
+func getInjectableYamlDocs(yamlDoc string, t *testing.T) [][]byte {
+	m := make(map[string]interface{})
+	if err := yaml.Unmarshal([]byte(yamlDoc), &m); err != nil {
+		t.Fatal(err)
+	}
+	switch m["kind"] {
+	case "Deployment":
+		return [][]byte{[]byte(yamlDoc)}
+	case "DeploymentConfig":
+		return [][]byte{[]byte(yamlDoc)}
+	case "DaemonSet":
+		return [][]byte{[]byte(yamlDoc)}
+	case "StatefulSet":
+		return [][]byte{[]byte(yamlDoc)}
+	case "Job":
+		return [][]byte{[]byte(yamlDoc)}
+	case "ReplicaSet":
+		return [][]byte{[]byte(yamlDoc)}
+	case "ReplicationController":
+		return [][]byte{[]byte(yamlDoc)}
+	case "CronJob":
+		return [][]byte{[]byte(yamlDoc)}
+	case "List":
+		// Split apart the list into separate yaml documents.
+		out := [][]byte{}
+		list := metav1.List{}
+		if err := yaml.Unmarshal([]byte(yamlDoc), &list); err != nil {
+			t.Fatal(err)
+		}
+		for _, i := range list.Items {
+			iout, err := yaml.Marshal(i)
+			if err != nil {
+				t.Fatal(err)
+			}
+			injectables := getInjectableYamlDocs(string(iout), t)
+			for _, injectable := range injectables {
+				out = append(out, injectable)
+			}
+		}
+		return out
+	default:
+		// No injectable parts.
+		return [][]byte{}
+	}
+}
+
+func toJSON(i interface{}, t *testing.T) []byte {
+	outputJSON, err := json.Marshal(i)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return prettyJSON(outputJSON, t)
+}
+
+func yamlToJSON(inputYAML []byte, t *testing.T) []byte {
+	// Convert to JSON
+	out, err := yaml.YAMLToJSON(inputYAML)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	return prettyJSON(out, t)
+}
+
+func prettyJSON(inputJSON []byte, t *testing.T) []byte {
+	// Pretty-print the JSON
+	var prettyBuffer bytes.Buffer
+	if err := json.Indent(&prettyBuffer, inputJSON, "", "  "); err != nil {
+		t.Fatalf(err.Error())
+	}
+	return prettyBuffer.Bytes()
+}
+
+func applyJSONPatch(input, patch []byte, t *testing.T) []byte {
+	p, err := jsonpatch.DecodePatch(patch)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	patchedJSON, err := p.Apply(input)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return prettyJSON(patchedJSON, t)
+}
+
+func jsonToDeployment(deploymentJSON []byte, t *testing.T) *extv1beta1.Deployment {
+	var deployment extv1beta1.Deployment
+	if err := json.Unmarshal(deploymentJSON, &deployment); err != nil {
+		t.Fatal(err)
+	}
+	return &deployment
+}
+
+func compareDeployments(got, want *extv1beta1.Deployment, t *testing.T) {
+	// Scrub unimportant fields that tend to differ.
+	annotations(got)["sidecar.istio.io/status"] = annotations(want)["sidecar.istio.io/status"]
+	gotIstioCerts := istioCerts(got)
+	wantIstioCerts := istioCerts(want)
+	if gotIstioCerts == nil || wantIstioCerts == nil {
+		if gotIstioCerts != nil || wantIstioCerts != nil {
```

Comment:
optional: It took me a moment to figure out what this was doing. The following might be clearer?

```golang
if (gotIstioCerts != nil) != (wantIstioCerts != nil) {}
```